src/interpreter/processor.rb

```diff
@@ -0,0 +1,212 @@
+Dir["#{__dir__}/processor/token_processors/*.rb"].each { |f| require_relative f }
+
+module Interpreter
+  class Processor
+    MAX_CALL_STACK_DEPTH = 1000
+
+    ############################################################################
+    # TokenProcessors consist of processing methods and any processing-specific
+    # helper methods.
+    #
+    # Processing:
+    # These methods take tokens and execute them. They may also read from the
+    # stack or additional tokens in the token stream. The stack should be clear
+    # after each call to a processing method.
+    #
+    # Any other common or helper methods will be in this file.
+    ############################################################################
+    include TokenProcessors
+
+    include BuiltIns
+
+    def initialize(lexer, options = {})
+      @lexer   = lexer
+      @options = options
+
+      @current_scope = Scope.new
+
+      # The current stack of tokens which have not been processed.
+      @stack = []
+    end
+
+    def execute
+      process
+    rescue Errors::BaseError => e
+      e.line_num = @lexer.line_num
+      raise
+    end
+
+    private
+
+    def process
+      raise Errors::CallStackTooDeep, MAX_CALL_STACK_DEPTH if caller.length > MAX_CALL_STACK_DEPTH
+
+      loop do
+        token = next_token
+        break if token.nil?
+
+        result = process_token token
+        return result if result.is_a? ReturnValue
+      end
+
+      nil
+    end
+
+    def next_token
+      token = peek_next_token
+      Util::Logger.debug Util::Options::DEBUG_2, 'RECEIVE: '.lred + (token ? "#{token} #{token.content}" : 'EOF')
+      @current_scope.advance
+      token
+    end
+
+    def peek_next_token
+      if @current_scope.type == Scope::TYPE_MAIN
+        token = @lexer.next_token
+        @current_scope.tokens << token unless token.nil?
+      end
+      @current_scope.current_token
+    end
+
+    def next_token_if(token_type)
+      next_token if peek_next_token&.type == token_type
+    end
+
+    # Accumulates tokens until the requested token type.
+    # If searching for a SCOPE_CLOSE: skips pairs of matching SCOPE_BEGINS and
+    # SCOPE_CLOSEs.
+    def next_tokens_until(token_type, options = { inclusive?: true })
+      [].tap do |tokens|
+        open_count = 0
+
+        loop do
+          peeked_token = peek_next_token
+
+          open_count += 1 if token_type == Token::SCOPE_CLOSE && peeked_token.type == Token::SCOPE_BEGIN
+
+          if peeked_token.type == token_type
+            break if open_count.zero?
+            open_count -= 1
+          end
+
+          tokens << next_token
+        end
+
+        tokens << next_token if options[:inclusive?]
+      end
+    end
+
+    def next_tokens_from_scope_body
+      next_token # discard scope open
+      body_tokens = next_tokens_until Token::SCOPE_CLOSE
+      body_tokens.pop # discard scope close
+      body_tokens
+    end
+
+    def process_token(token)
```

Comment:
Doc comment:

> Calls the associated processing method if the current token can lead to meaningful execution. Otherwise, stores the token in the stack to be processed later.